File: server/webapp/WEB-INF/rails.new/app/presenters/api_v1/config/pipeline_config_representer.rb
##########################################################################
# Copyright 2015 ThoughtWorks, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
##########################################################################

module ApiV1
  module Config
    class PipelineConfigRepresenter < ApiV1::BaseRepresenter
      alias_method :pipeline, :represented

      link :self do |opts|
        opts[:url_builder].apiv1_admin_pipeline_url(name: pipeline.name)
      end

      link :doc do |opts|
        'http://api.go.cd/#pipeline_config'
      end

      link :find do |opts|
        opts[:url_builder].apiv1_admin_pipeline_url(name: ':name')
      end

      property :label_template
      property :enable_pipeline_locking,
               getter: lambda { |options| self.isLock },
               setter: lambda { |val, options| val ? self.lockExplicitly : self.unlockExplicitly }

      property :name, case_insensitive_string: true
      property :template_name, as: :template, case_insensitive_string: true

      collection :parameters,
                 exec_context: :decorator,
                 decorator:    ApiV1::Config::ParamRepresenter,
                 expect_hash:  true,
                 class:        com.thoughtworks.go.config.ParamConfig

      collection :environment_variables,
                 exec_context: :decorator,
                 decorator:    ApiV1::Config::EnvironmentVariableRepresenter,
                 expect_hash:  true,
                 class:        com.thoughtworks.go.config.EnvironmentVariableConfig

      collection :materials,
                 exec_context: :decorator,
                 decorator:    ApiV1::Config::Materials::MaterialRepresenter,
                 expect_hash:  true,
                 class:        lambda { |fragment, *|
                   ApiV1::Config::Materials::MaterialRepresenter.get_material_type(fragment[:type]||fragment['type'])
                 }
      collection :stages,
                 exec_context: :decorator,
                 decorator:    ApiV1::Config::StageRepresenter,
                 expect_hash:  true,
                 class:        com.thoughtworks.go.config.StageConfig

      property :tracking_tool,
               exec_context: :decorator,
               decorator:    ApiV1::Config::TrackingTool::TrackingToolRepresenter,
               expect_hash:  true,
               class:        lambda { |object, *|
                 ApiV1::Config::TrackingTool::TrackingToolRepresenter.get_class(object[:type] || object['type'])
               }

      property :timer,
               decorator:  ApiV1::Config::TimerRepresenter,
               class:      com.thoughtworks.go.config.TimerConfig,
               skip_parse: SkipParseOnBlank

      property :errors,
               exec_context: :decorator,
               decorator:    ApiV1::Config::ErrorRepresenter,
               skip_parse:   true,
               skip_render:  lambda { |object, options| object.empty? }


      delegate :name, :name=, to: :pipeline
      def parameters
        pipeline.params
      end

      def parameters=(new_params)
        pipeline.params = new_params
      end

      def environment_variables
        pipeline.getVariables()
      end

      def environment_variables=(array_of_variables)
        pipeline.setVariables(EnvironmentVariablesConfig.new(array_of_variables))
      end

      def materials
        pipeline.materialConfigs()
      end

      def materials=(value)
        pipeline.materialConfigs().clear
        value.each { |material| pipeline.materialConfigs().add(material) }
      end

      def stages
        pipeline.getStages() if !pipeline.getStages().isEmpty
      end

      def stages=(value)
        pipeline.getStages().clear()
        value.each { |stage| pipeline.addStageWithoutValidityAssertion(stage) }
      end

      def tracking_tool
        if pipeline.getTrackingTool()
          pipeline.getTrackingTool()
        elsif pipeline.getMingleConfig().isDefined()
          pipeline.getMingleConfig()
        end

      end

      def tracking_tool=(value)
        if value.instance_of? com.thoughtworks.go.config.MingleConfig
          pipeline.setMingleConfig(value)
        elsif value.instance_of? com.thoughtworks.go.config.TrackingTool
          pipeline.setTrackingTool(value)
        end
      end

      def errors
        pipeline.errors.addAll(pipeline.materialConfigs.errors)
        pipeline.errors
      end

    end
  end
end

File: api/api-pipeline-config-v6/src/main/java/com/thoughtworks/go/apiv6/admin/pipelineconfig/representers/PipelineConfigRepresenter.java
/*
 * Copyright 2018 ThoughtWorks, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.thoughtworks.go.apiv6.admin.pipelineconfig.representers;

import com.thoughtworks.go.api.base.OutputListWriter;
import com.thoughtworks.go.api.base.OutputWriter;
import com.thoughtworks.go.api.representers.ErrorGetter;
import com.thoughtworks.go.api.representers.JsonReader;
import com.thoughtworks.go.apiv6.admin.pipelineconfig.representers.materials.MaterialRepresenter;
import com.thoughtworks.go.apiv6.admin.pipelineconfig.representers.trackingtool.TrackingToolRepresenter;
import com.thoughtworks.go.apiv6.shared.representers.EnvironmentVariableRepresenter;
import com.thoughtworks.go.apiv6.shared.representers.configorigin.ConfigRepoOriginRepresenter;
import com.thoughtworks.go.apiv6.shared.representers.configorigin.ConfigXmlOriginRepresenter;
import com.thoughtworks.go.apiv6.shared.representers.stages.StageRepresenter;
import com.thoughtworks.go.config.*;
import com.thoughtworks.go.config.exceptions.UnprocessableEntityException;
import com.thoughtworks.go.config.materials.MaterialConfigs;
import com.thoughtworks.go.config.remote.ConfigOrigin;
import com.thoughtworks.go.config.remote.FileConfigOrigin;
import com.thoughtworks.go.config.remote.RepoConfigOrigin;
import com.thoughtworks.go.spark.Routes;
import org.bouncycastle.crypto.InvalidCipherTextException;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Consumer;

public class PipelineConfigRepresenter {

    public static void toJSON(OutputWriter jsonWriter, PipelineConfig pipelineConfig) {
        jsonWriter.addLinks(linksWriter -> linksWriter
                .addLink("self", Routes.PipelineConfig.name(pipelineConfig.getName().toString()))
                .addAbsoluteLink("doc", Routes.PipelineConfig.DOC)
                .addLink("find", Routes.PipelineConfig.find()));
        if (!pipelineConfig.getAllErrors().isEmpty()) {
            jsonWriter.addChild("errors", errorWriter -> {
                HashMap<String, String> errorMapping = new HashMap<>();
                errorMapping.put("labelTemplate", "label_template");
                errorMapping.put("params", "parameters");
                errorMapping.put("variables", "environment_variables");
                errorMapping.put("trackingTool", "tracking_tool");

                new ErrorGetter(errorMapping).toJSON(errorWriter, pipelineConfig);
            });
        }

        jsonWriter.add("label_template", pipelineConfig.getLabelTemplate());
        jsonWriter.add("lock_behavior", pipelineConfig.getLockBehavior());
        jsonWriter.add("name", pipelineConfig.name());
        jsonWriter.add("template", pipelineConfig.getTemplateName());
        writeOrigin(jsonWriter, pipelineConfig.getOrigin());
        jsonWriter.addChildList("parameters", getParams(pipelineConfig.getParams()));
        jsonWriter.addChildList("environment_variables", envVarsWriter -> EnvironmentVariableRepresenter.toJSON(envVarsWriter, pipelineConfig.getVariables()));
        jsonWriter.addChildList("materials", getMaterials(pipelineConfig.materialConfigs()));
        if (pipelineConfig.getStages().isEmpty()) {
            jsonWriter.renderNull("stages");
        } else {
            jsonWriter.addChildList("stages", getStages(pipelineConfig));
        }
        if ((pipelineConfig.getTrackingTool() != null && pipelineConfig.getTrackingTool().isDefined()) || pipelineConfig.getMingleConfig().isDefined()) {
            jsonWriter.addChild("tracking_tool", trackingToolWriter -> TrackingToolRepresenter.toJSON(trackingToolWriter, pipelineConfig));
        } else {
            jsonWriter.renderNull("tracking_tool");
        }
        if (pipelineConfig.getTimer() == null) {
            jsonWriter.renderNull("timer");
        } else {
            jsonWriter.addChild("timer", timerWriter -> TimerRepresenter.toJSON(timerWriter, pipelineConfig.getTimer()));
        }
    }

    private static Consumer<OutputListWriter> getMaterials(MaterialConfigs materialConfigs) {
        return materialsWriter -> {
            materialConfigs.forEach(materialConfig -> {
                materialsWriter.addChild(materialWriter -> MaterialRepresenter.toJSON(materialWriter, materialConfig));
            });
        };
    }

    private static Consumer<OutputListWriter> getParams(ParamsConfig params) {
        return paramsWriter -> {
            params.forEach(param -> {
                paramsWriter.addChild(paramWriter -> ParamRepresenter.toJSON(paramWriter, param));
            });
        };
    }

    private static Consumer<OutputListWriter> getStages(PipelineConfig pipelineConfig) {
        return stagesWriter -> {
            pipelineConfig.getStages().forEach(stage -> {
                stagesWriter.addChild(stageWriter -> StageRepresenter.toJSON(stageWriter, stage));
            });
        };
    }

    private static void writeOrigin(OutputWriter jsonWriter, ConfigOrigin origin) {
        if (origin instanceof FileConfigOrigin) {
            jsonWriter.addChild("origin", originWriter -> ConfigXmlOriginRepresenter.toJSON(originWriter, (FileConfigOrigin) origin));
        } else {
            jsonWriter.addChild("origin", originWriter -> ConfigRepoOriginRepresenter.toJSON(originWriter, (RepoConfigOrigin) origin));
        }
    }

    public static PipelineConfig fromJSON(JsonReader jsonReader, Map<String, Object> options) {
        if (jsonReader == null) {
            return null;
        }
        PipelineConfig pipelineConfig = new PipelineConfig();
        jsonReader.readStringIfPresent("label_template", pipelineConfig::setLabelTemplate);
        jsonReader.readStringIfPresent("lock_behavior", pipelineConfig::setLockBehaviorIfNecessary);
        jsonReader.readStringIfPresent("name", pipelineConfig::setName);
        jsonReader.readCaseInsensitiveStringIfPresent("template", pipelineConfig::setTemplateName);
        pipelineConfig.setOrigin(new FileConfigOrigin());
        setParameters(jsonReader, pipelineConfig);
        setEnvironmentVariables(jsonReader, pipelineConfig);
        setMaterials(jsonReader, pipelineConfig, options);
        setStages(jsonReader, pipelineConfig);
        setTrackingTool(jsonReader, pipelineConfig);
        setTimer(jsonReader, pipelineConfig);
        return pipelineConfig;
    }

    private static void setTimer(JsonReader jsonReader, PipelineConfig pipelineConfig) {
        if (jsonReader.hasJsonObject("timer")) {
            TimerConfig timer = TimerRepresenter.fromJSON(jsonReader.readJsonObject("timer"));
            pipelineConfig.setTimer(timer);
        }
    }

    private static void setTrackingTool(JsonReader jsonReader, PipelineConfig pipelineConfig) {
        if (jsonReader.hasJsonObject("tracking_tool")) {
            Object trackingTool = TrackingToolRepresenter.fromJSON(jsonReader.readJsonObject("tracking_tool"));
            if (trackingTool instanceof MingleConfig) {
                pipelineConfig.setMingleConfig((MingleConfig) trackingTool);
            }
            else if (trackingTool instanceof TrackingTool) {
                pipelineConfig.setTrackingTool((TrackingTool) trackingTool);
            }
        }
    }

    private static void setStages(JsonReader jsonReader, PipelineConfig pipelineConfig) {
        pipelineConfig.getStages().clear();
        jsonReader.readArrayIfPresent("stages", stages -> {
            stages.forEach(stage -> {
                pipelineConfig.addStageWithoutValidityAssertion(StageRepresenter.fromJSON(new JsonReader(stage.getAsJsonObject())));
            });
        });
    }

    private static void setMaterials(JsonReader jsonReader, PipelineConfig pipelineConfig, Map<String, Object> options) {
        MaterialConfigs materialConfigs = new MaterialConfigs();
        jsonReader.readArrayIfPresent("materials", materials -> {
            materials.forEach(material -> {
                materialConfigs.add(MaterialRepresenter.fromJSON(new JsonReader(material.getAsJsonObject()), options));
            });
        });
        pipelineConfig.setMaterialConfigs(materialConfigs);
    }

    private static void setParameters(JsonReader jsonReader, PipelineConfig pipelineConfig) {
        ParamsConfig paramConfigs = new ParamsConfig();
        jsonReader.readArrayIfPresent("parameters", params -> {
            params.forEach(param -> {
                paramConfigs.add(ParamRepresenter.fromJSON(new JsonReader(param.getAsJsonObject())));
            });
        });
        pipelineConfig.setParams(paramConfigs);
    }

    private static void setEnvironmentVariables(JsonReader jsonReader, PipelineConfig pipelineConfig) {
        EnvironmentVariablesConfig environmentVariableConfigs = new EnvironmentVariablesConfig();
        jsonReader.readArrayIfPresent("environment_variables", variables -> {
            variables.forEach(variable -> {
                try {
                    environmentVariableConfigs.add(EnvironmentVariableRepresenter.fromJSON(new JsonReader(variable.getAsJsonObject())));
                } catch (InvalidCipherTextException e) {
                    e.printStackTrace();
                }
            });
        });
        pipelineConfig.setVariables(environmentVariableConfigs);
    }
}
